src/Extensions/Logging/Internal/Replayable/ReplayableActivityExtensions.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.Omex.Extensions.Logging.Replayable;
+
+namespace Microsoft.Omex.Extensions.Logging.Internal.Replayable
+{
+	internal static class ReplayableActivityExtensions
+	{
+		private const string ReplayableLogKey = "ReplayableLogKey";
+
+		private static ConcurrentQueue<LogMessageInformation>? GetReplayableLogsInternal(this Activity activity) =>
+			activity.GetCustomProperty(ReplayableLogKey) as ConcurrentQueue<LogMessageInformation>;
+
+		public static IEnumerable<LogMessageInformation> GetReplayableLogs(this Activity activity) =>
+			activity.GetReplayableLogsInternal() ?? Enumerable.Empty<LogMessageInformation>();
+
+		public static void AddReplayLog(this Activity activity, LogMessageInformation logEvent, uint maxReplayedEventsPerActivity)
+		{
+			ConcurrentQueue<LogMessageInformation>? queue = activity.GetReplayableLogsInternal();
+			if (queue == null)
+			{
+				Interlocked.CompareExchange(ref queue, new ConcurrentQueue<LogMessageInformation>(), null);
```

Comment:
>queue [](start = 36, length = 5)

Just thinking if we should not read the value of queue again here. If the CAS fails because another thread changed the value, will queue contain the new value or we have to read it again with queue = activity.GetReplayableLogsInternal();?


The pattern is usually to do a while loop until the condition is satisfied, i.e. do an interlocked read, then do a swap and if it fails , do an interlocked read again. The interlocked operations are flushing the memory as well. 

Is that a method called so often? A simple lock might be much easier to implement, but would be a bottleneck if that's called more often.

Reply:
To be honest I don't see a good solution here.
I've considered something like this:
```
ConcurrentQueue<LogMessageInformation>? queue;
while ((queue = activity.GetReplayableLogsInternal()) == null)
{
	activity.SetCustomProperty(ReplayableLogKey, new ConcurrentQueue<LogMessageInformation>());
}
```
But it would end up with the same result, we could still have one query be replaced by another right after creation.

I also don't see a good way of adding lock here, the only object that could be used as a lock is the `activity` itself but that's dangerous since it's public and could be used in some other lock. 

I also think that it's not the most critical case here since in the worst case we are losing a couple of messages to replay, it's not great but we can allow it considering alternatives.

Besides that, I really hope that I'll eventually manage to convince everybody that log replay is not suitable for microservices and we should use variable report level instead, so we would be able to delete reply code completely.